src/ParcelRegistry.Importer.Grb/Importer.cs

```diff
@@ -0,0 +1,126 @@
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Be.Vlaanderen.Basisregisters.AggregateSource;
+    using Be.Vlaanderen.Basisregisters.Utilities.HexByteConvertor;
+    using Infrastructure;
+    using Infrastructure.Download;
+    using MediatR;
+    using Microsoft.Extensions.Hosting;
+    using Parcel.Exceptions;
+
+    public class Importer : BackgroundService
+    {
+        private readonly IMediator _mediator;
+        private readonly IUniqueParcelPlanProxy _uniqueParcelPlanProxy;
+        private readonly IZipArchiveProcessor _zipArchiveProcessor;
+        private readonly IRequestMapper _requestMapper;
+        private readonly IImporterContext _importerContext;
+        private readonly INotificationService _notificationService;
+
+        public Importer(
+            IMediator mediator,
+            IUniqueParcelPlanProxy uniqueParcelPlanProxy,
+            IZipArchiveProcessor zipArchiveProcessor,
+            IRequestMapper requestMapper,
+            IImporterContext importerContext,
+            INotificationService notificationService)
+        {
+            _mediator = mediator;
+            _uniqueParcelPlanProxy = uniqueParcelPlanProxy;
+            _zipArchiveProcessor = zipArchiveProcessor;
+            _requestMapper = requestMapper;
+            _importerContext = importerContext;
+            _notificationService = notificationService;
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            try
+            {
+                var lastRun = await _importerContext.GetLatestRunHistory();
+                RunHistory currentRun;
+                if (lastRun.Completed)
+                {
+                    var maxDate = await _uniqueParcelPlanProxy.GetMaxDate();
+                    currentRun = await _importerContext.AddRunHistory(lastRun.ToDate, maxDate);
+                }
+                else
+                {
+                    currentRun = lastRun;
+                }
+
+                var zipArchive = await _uniqueParcelPlanProxy.Download(currentRun.FromDate, currentRun.ToDate);
+
+                var files = _zipArchiveProcessor.Open(zipArchive);
+
+                var parcelsRequests = _requestMapper.Map(files);
+
+                var groupedParcels = parcelsRequests
+                    .OrderBy(x => x.GrbParcel.Version)
+                    .GroupBy(y => y.GrbParcel.GrbCaPaKey);
+
+                foreach (var parcelRequest in groupedParcels.SelectMany(x => x))
+                {
+                    try
+                    {
+                        if (await _importerContext.ProcessedRequestExists(parcelRequest.GetSHA256()))
+                        {
+                            continue;
+                        }
+
+                        await _mediator.Send(parcelRequest, stoppingToken);
+
+                        await _importerContext.AddProcessedRequest(parcelRequest.GetSHA256());
```

Comment:
If there's a crash between these lines, then retry will try and process this request again.
I think it's safer to first `AddProcessedRequest` and then do `mediator.send`, but on Exception, remove the hash (CancellationToken.None).

Reply:
Also extract the hash in a variable above the `try{`, you recalculate it twice now, and a third if you also do the above implementation.